# Prova di esame (laboratorio)

La prova avrà una durata massima di due ore e porterà a una valutazione massima di 21 punti. Il voto finale si otterrà per somma del punteggio del test di teoria e della prova di laboratorio. 

La prova consiste in un insieme di quiz proposti attraverso un Jupyter Notebook (non sono ammessi altri strumenti software) da svolgere sui computer del laboratorio (non sono ammessi altri strumenti hardware). Durante la prova si può consultare il libro di testo, le slide usate a lezione, un manuale python (approvato dal docente) oppure le guide in linea suggerite da Jupyter Notebook. Non sono ammessi altri strumenti di consultazione. 

## Traccia

**Consumer Price Index in Argentina**

Serie temporale dell'indice dei prezzi al consumo (CPI) in Argentina (indice con 1969 = 1)

*Formato*

Il dataset contiene 80 osservazioni su 2 variabili oltre all'indice di riga.

1. Quartile (es. 1970 = I trimestre 1970; 1970.25 = II trimestre 1970; 1970.5 = III trimestre 1970; 1970.75 = IV trimestre 1970)
2. Indice dei prezzi al consumo (es. 4.29 = i prezzi al consumo sono 4.29 volte quelli del 1969)


**Prima della prova**

Specificare la informazioni richieste nelle celle sottostanti e salvare il notebook con il proprio nome.

**Durante la prova**

Nelle celle markdown, cancellare il testo "YOUR ANSWER HERE" con la propria risposta. Nelle celle codice, rimuovere la linea di codice
```python
raise NotImplementedError()
```
e sostituirla con la propria implementazione.
Dopo ogni cella codice che deve contenere la risposta al quesito, esiste una cella codice con delle *asserzioni*, ossia delle istruzioni che verificano la correttezza della propria soluzione. (Queste asserzioni saranno usate per valutare la correttezza della risposta data e assegnare un punteggio.) E' possibile eseguire queste celle per verificare che la soluzione proposta sia corretta.


**Dopo la prova**

Salvare il file dopo aver terminato. Consegnare l'elaborato utilizzando la piattaforma ADA. 

### Caricamento dati (3 pt)

Utilizzando il modulo csv, caricare i dati in un dizionario di nome `data` che associ a ogni quartile il corrispondente indice dei prezzi. Non caricare la prima riga, che contiene le intestazioni di colonna. Assicurarsi che chiavi e valori siano numerici.

In [2]:
import csv

with open('ArgentinaCPI.csv') as arg:
    data = dict(map(float , elem[1 :]) for elem in list(csv.reader(arg))[1 :])

print(list(data.items())[0:10])

[(1970.0, 1.01), (1970.25, 1.06), (1970.5, 1.13), (1970.75, 1.24), (1971.0, 1.33), (1971.25, 1.45), (1971.5, 1.64), (1971.75, 1.77), (1972.0, 2.24), (1972.25, 2.58)]


In [3]:
assert data[1970.25] > 1

In [4]:
assert len(data) == 80

### Numpy (3 pt)

Definire un array numpy denominato `cpi` costituito da tutti i *valori* contenuti nel dizionario `data`

In [5]:
import numpy as np

cpi = np.array(list(data.values()))

cpi[0:10]

array([1.01, 1.06, 1.13, 1.24, 1.33, 1.45, 1.64, 1.77, 2.24, 2.58])

In [6]:
assert 2 >= cpi[0] >= 1

In [7]:
assert cpi.shape == (80,)

### Funzioni (3 pt)

Definire una funzione denominata `diag` con argomento `X` che si assume essere un array bidimensionale (matrice). La funzione restituisce un array corrispondente alla diagonale di `X`, ossia a tutti gli elementi con stesso indice di riga e colonna. La funzione `diag` deve restituire la diagonale anche quando `X` è una matrica rettangolare (cioé quando il numero di righe e colonne non coincidono).

In [8]:
def diag(X):
    
    return np.array([X[i][i] for i in range(len(X) if len(X) <= len(X[0]) else len(X[0]))])

    
sample = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
])

print(diag(sample))

[1 5 9]


In [9]:
assert all(diag(sample) == np.array([1, 5, 9]))

In [10]:
assert all(diag(sample[:,0:2]) == np.array([1, 5]))

In [11]:
assert all(diag(sample[1:,:]) == np.array([4, 8]))

### Reshaping (3 pt)

Ridimensionare l'array `cpi` in una matrice di 4 colonne per 20 righe denominata `cpi_mx`. Applicare la funzione `diag` alla matrice per estrarre gli elementi diagonali e memorizzarli nell'array `cpi_d`.

In [12]:
cpi_mx = cpi.reshape(20 , 4)
cpi_d = diag(cpi_mx)

print(cpi_mx[:5,:])
print(cpi_d)

[[1.01 1.06 1.13 1.24]
 [1.33 1.45 1.64 1.77]
 [2.24 2.58 2.92 3.2 ]
 [4.26 4.89 4.94 5.15]
 [5.19 5.57 6.   6.92]]
[1.01 1.45 2.92 5.15]


In [13]:
assert cpi_mx.shape == (20, 4)

In [14]:
assert all(cpi_d == np.array([1.01, 1.45, 2.92, 5.15]))

### Comprensioni (3 pt)

Definire una lista denominata `ratios` dove ciascun elemento corrisponde al rapporto di un valore nell'array `cpi` con quello di quattro posizioni precedenti (per esempio, `ratios[0] == cpi[4]/cpi[0]`). Utilizzare le comprensioni di lista per generare la lista. 

In [15]:
ratios = [cpi[i] / cpi[i - 4] for i in range(4 , len(cpi))]

print(ratios[0:10])

[1.316831683168317, 1.3679245283018866, 1.4513274336283186, 1.4274193548387097, 1.6842105263157896, 1.7793103448275862, 1.7804878048780488, 1.807909604519774, 1.901785714285714, 1.8953488372093021]


In [16]:
assert type(ratios)==list

In [17]:
assert len(ratios)==76

In [18]:
assert all(r > 1 for r in ratios)

### Formattazione di stringhe (3 pt)

Utilizzando il metodo `format` della classe `str` definire una lista di stringhe denominata `cpi_s` i cui elementi sono rappresentazioni in stringa degli elementi corrispondenti di `cpi` formattati in 13 caratteri con due cifre decimali ed eventuali zeri iniziali (p.e. `1.316831683168317` --> `'0000000001.32'`

In [19]:
cpi_s = ['{0:013.2f}'.format(elem) for elem in cpi]

cpi_s[70:]

['0022297936.00',
 '0032220516.00',
 '0041661128.00',
 '0064616412.00',
 '0116430000.00',
 '0155770000.00',
 '0203130000.00',
 '0679960000.00',
 '4537300000.00',
 '6611700000.00']

In [20]:
assert all(len(s)==13 for s in cpi_s)

In [21]:
import re
assert all(re.fullmatch(r'\d+\.\d\d', s) for s in cpi_s)

In [22]:
assert all(re.fullmatch(r'^\d.*',s) for s in cpi_s)

### Programmazione funzionale

Utilizzando la funzione `filter` e le espressioni lambda, generare una lista denominata `filtered_cpi` che contiene le stringhe di `cpi_s` che soddisfino i seguenti vincoli:
1. La cifra corrispondente all'unità deve essere pari;
2. Il numero rappresentato deve essere intero (cioé con parte decimale uguale a 0) 

In [23]:
import re 

filtered_cpi = list(filter(lambda x : re.fullmatch(r'\d{9}[02468].0{2}' , x) , cpi_s))

filtered_cpi

['0000000006.00',
 '0000000744.00',
 '0001196314.00',
 '0002180880.00',
 '0004484544.00',
 '0007194400.00',
 '0009834494.00',
 '0011798442.00',
 '0014267856.00',
 '0016727634.00',
 '0022297936.00',
 '0032220516.00',
 '0041661128.00',
 '0064616412.00',
 '0116430000.00',
 '0155770000.00',
 '0203130000.00',
 '0679960000.00',
 '4537300000.00',
 '6611700000.00']

In [24]:
assert type(filtered_cpi[0])==str

In [25]:
assert all(re.fullmatch(r'.*\.00', s) for s in filtered_cpi)

In [26]:
assert all(re.fullmatch(r'\d{9}[02468].*', s) for s in filtered_cpi)